### PR TITLE
Add Azure Provisioner

### PR DESCRIPTION
**Pull request by zechenbit**
Signed-off-by: Ze Chen <zechenbit@gmail.com>

- [x] I have raised an issue to propose this change.

## Description

The following up changes in inlets-operator for #47 

## How Has This Been Tested?

**Create**
```
kubectl apply -f ./artifacts/crds/
kubectl create secret generic inlets-access-key --from-file=inlets-access-key=~/Downloads/client_credentials.json
helm repo add inlets https://inlets.github.io/inlets-operator/
helm repo update
helm upgrade inlets-operator --install inlets/inlets-operator \
  --set provider=azure,region=eastus,subscriptionID=<Azure Subscription ID>

kubectl run nginx-1 --image=nginx --port=80 --restart=Always
kubectl expose deployment nginx-1 --port=80 --type=LoadBalancer
kubectl get svc -w
```

Output:
```
deployment.apps/nginx-1 created
service/nginx-1 exposed
done
NAME         TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)        AGE
kubernetes   ClusterIP      10.43.0.1       <none>        443/TCP        50s
nginx-1      LoadBalancer   10.43.178.247   <pending>     80:31706/TCP   10s
nginx-1      LoadBalancer   10.43.178.247   172.22.0.2    80:31706/TCP   14s
nginx-1      LoadBalancer   10.43.178.247   172.22.0.2,137.135.126.179   80:31706/TCP   2m46s
nginx-1      LoadBalancer   10.43.178.247   137.135.126.179,137.135.126.179   80:31706/TCP   2m46s
nginx-1      LoadBalancer   10.43.178.247   172.22.0.2,137.135.126.179        80:31706/TCP   2m46s
```

**Delete**
```
➜  inlets-operator git:(provision-azure) ✗ kubectl delete service nginx-1
service "nginx-1" deleted
```

## How are existing users impacted? What migration steps/scripts do we need?
No impact, this is new feature.

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/inlets/inlets/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

### CODE REVIEW

**Review by Waterdrips — APPROVED** [review]
Lgtm

**Comment by alexellis** [comment]
Hi @zechenbit, this needs a rebase on the README now after removing the "kubectl apply -f artifacts/" approach to installing the operator.

File: controller.go [hunk 1]
```diff
@@ -617,6 +617,32 @@ func getHostConfig(c *Controller, tunnel *inletsv1alpha1.Tunnel) provision.Basic
...
+
+		inletsPort := inletsOSSControlPort
+
+		if c.infraConfig.UsePro() {
+			inletsPort = inletsPROControlPort
+		}
+		host = provision.BasicHost{
+			Name:     tunnel.Name,
+			OS:       "Additional.imageOffer",
+			Plan:     "Standard_B1ls",
```
Comment (alexellis) [inline controller.go:634]: Was that the same plan as in the inletsctl version? Is there a cheaper option, or is this the best one for now?

File: main.go [hunk 2]
```diff
@@ -83,13 +84,14 @@ func main() {
...
 
-	flag.StringVar(&infra.Provider, "provider", "packet", "Your infrastructure provider - 'packet', 'digitalocean', 'scaleway', 'civo', 'gce', 'linode' or 'ec2'")
+	flag.StringVar(&infra.Provider, "provider", "packet", "Your infrastructure provider - 'packet', 'digitalocean', 'scaleway', 'civo', 'gce', 'linode', 'azure' or 'ec2'")
 	flag.StringVar(&infra.Region, "region", "", "The region to provision hosts into")
 	flag.StringVar(&infra.Zone, "zone", "us-central1-a", "The zone where the exit node is to be provisioned")
 	flag.StringVar(&infra.AccessKey, "access-key", "", "The access key for your infrastructure provider")
 	flag.StringVar(&infra.AccessKeyFile, "access-key-file", "", "Read the access key for your infrastructure provider from a file (recommended)")
 	flag.StringVar(&infra.SecretKey, "secret-key", "", "The secret key if using Scaleway or EC2 as the provider")
 	flag.StringVar(&infra.SecretKeyFile, "secret-key-file", "", "Read the access key for your infrastructure provider from a file (recommended)")
+	flag.StringVar(&infra.SubscriptionID, "subscription-id", "", "The Azure Subscription ID")
```
Comment (alexellis) [inline main.go:94]: If subscription-id is required for Azure, then can you make it part of the start-up validation and add a unit test also?

**Review by alexellis — CHANGES_REQUESTED** [review]
Changes look good, I would like to see a bit of additional validation before merging.

**Comment by alexellis** [comment]
https://github.com/inlets/inlets-operator/blob/master/validate.go
https://github.com/inlets/inlets-operator/blob/master/validate_test.go

File: controller.go (hunk 1 above)
Comment (zechenbit) [reply controller.go:634]: It is same as what we have in inletsctl. It is the cheapest plan: https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-b-series-burstable

File: main.go (hunk 2 above)
Comment (zechenbit) [reply main.go:94]: Added the validation and unit test.

File: validate_test.go [hunk 3]
```diff
@@ -55,6 +55,19 @@ func Test_validateFlags_GCE_ProjectID(t *testing.T) {
 	}
 }
 
+func Test_validateFlags_Azure_SubscriptionID(t *testing.T) {
```
Comment (alexellis) [inline validate_test.go:58]: There should be a positive and negative test to prove that the code works. Testing only the negative case is 50% of what we need for this. Thanks for adding it.

**Review by alexellis — APPROVED** [review]
LGTM, thank you

**Pull request merged**